networkx/algorithms/euler.py

```diff
@@ -246,16 +246,25 @@ def has_eulerian_path(G):
     eulerian_path
     """
     if G.is_directed():
-        ins = G.in_degree
-        outs = G.out_degree
-        semibalanced_ins = sum(ins(v) - outs(v) == 1 for v in G)
-        semibalanced_outs = sum(outs(v) - ins(v) == 1 for v in G)
+        H = G.copy()
+        nodes_remove = [v for v in H if H.in_degree(v) == 0 and H.out_degree(v) == 0]
+        H.remove_nodes_from(nodes_remove)
```

Comment:
We might be able to save a "copy" operation when there aren't any isolated nodes.
Something like:
```
        nodes_remove = [v for v in G if ins[v] == 0 and outs[v] == 0]
        if nodes_remove:
            G = G.copy()
            G.remove_nodes_from(nodes_remove)
```
and then shift references to H to be references to G.

Note that ```ins[v]``` is faster than ```ins(v)``` when there is only one node to look up.
Could be done a few lines down too... probably not a big deal...